R/utils-spatial-filter.R

```diff
@@ -0,0 +1,145 @@
+#'  [sf::st_union()]. If an `sfc` object has MULTIPOLYGON geometry, the features
+#'  are unioned before being cast to POLYGON geometry with [sf::st_cast()]. All
+#'  geometries are checked for validity before conversion.
+#'
+#' @returns [prepare_spatial_filter()] returns a named list with the
+#'   geometryType, geometry (as ESRI JSON), and spatial relation predicate.
+#'
+#' @export
+#' @rdname spatial_filter
+#' @keywords internal
+prepare_spatial_filter <- function(
+    filter_geom,
+    crs,
+    predicate,
+    error_call = rlang::caller_env()
+) {
+
+  check_inherits_any(
+    filter_geom,
+    class = c("sfc", "sfg", "bbox"),
+    call = error_call
+  )
+
+  # NOTE: CRS cannot be missing
+  if (inherits(filter_geom, "bbox")) {
+    filter_geom <- sf::st_as_sfc(filter_geom)
+  } else if (any(!sf::st_is_valid(filter_geom))) {
+    filter_geom <- sf::st_make_valid(filter_geom)
+  }
+
+  # FIXME: Unsure how to handle sfg inputs w/o checking CRS
+  if (inherits(filter_geom, "sfg")) {
+    filter_crs <- crs
+  } else {
+    filter_crs <- sf::st_crs(filter_geom)
+
+    if (is.na(filter_crs)) {
+      filter_crs <- crs
+    }
+  }
+
+  # if an sfc_multipolygon we union and cast to polygon
+  # related issue: https://github.com/R-ArcGIS/arcgislayers/issues/4
+  if (inherits(filter_geom, "sfc_MULTIPOLYGON")) {
+    cli::cli_inform(
+      c(
+        "!" = "{.arg filter_geom} cannot be a {.val MULTIPOLYGON} geometry.",
+        "i" = "Using {.fn sf::st_union} and {.fn sf::st_cast} to create a
+        {.val POLYGON} for {.arg filter_geom}."
+      ),
+      call = error_call
+    )
+    filter_geom <- sf::st_union(filter_geom)
+    filter_geom <- sf::st_cast(filter_geom, to = "POLYGON")
+  } else if (inherits(filter_geom, "MULTIPOLYGON")) {
+    filter_geom <- sf::st_cast(filter_geom, "POLYGON")
+  }
+
+  # if its an sfc object it must be length one
+  if (inherits(filter_geom, "sfc")) {
+    if (length(filter_geom) > 1) {
+      filter_geom <- sf::st_union(filter_geom)
+    }
+    # extract the sfg object which is used to write Esri json
+    filter_geom <- filter_geom[[1]]
+  }
+
+  list(
+    geometryType = arcgisutils::determine_esri_geo_type(filter_geom),
+    geometry = arcgisutils::as_esri_geometry(filter_geom, crs = filter_crs),
+    spatialRel = match_spatial_rel(predicate, error_call = error_call)
+    # TODO is `inSR` needed if the CRS is specified in the geometry???
+  )
+}
+
+#' @description
+#' [match_spatial_rel()] takes a scalar character vector with a predicate name
+#' to a type of ESRI spatial relation.
+#'
+#' @returns [match_spatial_rel()] returns one of the following spatial binary predicates:
+#'
+#' - esriSpatialRelIntersects
+#' - esriSpatialRelContains
+#' - esriSpatialRelCrosses
+#' - esriSpatialRelOverlaps
+#' - esriSpatialRelTouches
+#' - esriSpatialRelWithin
+#'
+#' @export
+#' @rdname spatial_filter
+match_spatial_rel <- function(predicate, error_call = rlang::caller_env()) {
```

Comment:
I'm still not convinced that you need to have this as an argument lol but its not a hill I'll die on.

Reply:
It is basically habit at this point more than some well-reasoned argument. But, basically, I read the [rlang docs around this topic](https://rlang.r-lib.org/reference/topic-error-call.html) a couple years back and started adding `call` or `error_call` to any function that includes any input checks. 🤷‍♂️